assets/app/scripts/controllers/edit/buildConfig.js

```diff
@@ -0,0 +1,307 @@
+'use strict';
+
+/**
+ * @ngdoc function
+ * @name openshiftConsole.controller:EditBuildConfigController
+ * @description
+ * Controller of the openshiftConsole
+ */
+angular.module('openshiftConsole')
+  .controller('EditBuildConfigController', function ($scope, $routeParams, DataService, ProjectsService, $filter, ApplicationGenerator, Navigate, $location, AlertMessageService) {
+
+    $scope.projectName = $routeParams.project;
+    $scope.buildConfig = null;
+    $scope.alerts = {};
+    $scope.emptyMessage = "Loading...";
+    $scope.options = {};
+
+    $scope.buildFromTypes = [
+      {
+        "id": "ImageStreamTag",
+        "title": "Image Stream Tag"
+      },
+      {
+        "id": "DockerImage",
+        "title": "Docker Image Link"
+      },
+    ];
+
+    $scope.breadcrumbs = [
+      {
+        title: "Builds",
+        link: "project/" + $routeParams.project + "/browse/builds"
+      },
+      {
+        title: $routeParams.buildconfig,
+        link: "project/" + $routeParams.project + "/browse/builds/" + $routeParams.buildconfig
+      },
+      {
+        title: "Edit"
+      }
+    ];
+
+    $scope.buildFrom = {
+      projects: [],
+      imageStreams: [],
+      tags: {},
+    }
+    $scope.pushTo = {
+      projects: [],
+      imageStreams: [],
+      tags: {},
+    }
+
+    $scope.triggers = {
+      "webhook": false,
+      "imageChange": false,
+      "configChange": false
+    };
+
+    var watches = [];
+
+    ProjectsService
+      .get($routeParams.project)
+      .then(_.spread(function(project, context) {
+        $scope.project = project;
+        DataService.get("buildconfigs", $routeParams.buildconfig, context).then(
+          // success
+          function(buildConfig) {
+            $scope.buildConfig = buildConfig;
+            $scope.updatedBuildConfig = angular.copy($scope.buildConfig);
+            $scope.buildStrategy = $filter('buildStrategy')($scope.updatedBuildConfig);
+            $scope.envVars = $filter('envVarsPair')($scope.buildStrategy.env);
```

Comment:
`env` could be undefined since it is `omitempty`, which would cause runtime errors later when we pass it to the `osc-key-values` directive.